### PR TITLE
Basic authentication with ldap support

### PR DESCRIPTION
**Pull request by dpavlov**
Update BasicAuthenticationFilter to work with Active Directory service

### CODE REVIEW

**Comment by yahoocla** [comment]
Thank you for submitting this pull request, however I do not see a valid CLA on file for you. Before we can merge this request please visit https://yahoocla.herokuapp.com/ and agree to the terms. Thanks! :smile:

**Comment by patelh** [comment]
please sign CLA and then close PR and reopen PR.

**Comment by patelh** [comment]
@dpavlov Can you add how to use ldap to the documentation/readme so others know how to configure and use it.

File: app/controllers/BasicAuthenticationFilter.scala [hunk 1]
```diff
@@ -20,33 +20,80 @@ import play.api.mvc.Results.Unauthorized
...
 
     authorizedByHeader || authorizedByCookie
   }
 
-  private def addCookie(result: Future[Result]) =
-    result.map(_.withCookies(cookie))
+  private def isAuthorized(requestHeader: RequestHeader, authenticator: BasicAuthenticationFilterLdapAuthenticator, credentials: (String, String)) = {
+    val (username, password) = credentials
+    val expectedCookie = cookieValue(username, Set(password))
+    lazy val authorizedByCookie =
```
[the remark this answers is not in the document]
Comment (dpavlov) [reply app/controllers/BasicAuthenticationFilter.scala]: yes you are right. there is no reason to be lazy here. I will fix it

File: app/controllers/BasicAuthenticationFilter.scala [hunk 2]
```diff
@@ -20,33 +20,80 @@ import play.api.mvc.Results.Unauthorized
...
+      requestHeader.cookies.get(COOKIE_NAME).exists(_.value == expectedCookie)
+
+    authorizedByCookie || {
+      val connection = ldapConnectionPool.getConnection
+      try {
+        findUserDN(authenticator.searchBaseDN, authenticator.searchFilter, username, connection) match {
+          case None => false
+          case Some(userDN) => Try(connection.bind(userDN, password)).isSuccess
```
[the remark this answers is not in the document]
Comment (dpavlov) [reply app/controllers/BasicAuthenticationFilter.scala]: Done

File: app/controllers/BasicAuthenticationFilter.scala [hunk 3]
```diff
@@ -112,33 +213,62 @@ object BasicAuthenticationFilterConfiguration {
...
+      val password = string("ldap.password").getOrElse("")
+
+      val searchDN = string("ldap.search-base-dn").getOrElse("")
+      val searchFilter = string("ldap.search-filter").getOrElse("")
+      val connectionPoolSize = int("ldap.connection-pool-size").getOrElse(10)
+      val sslEnabled = boolean("ldap.ssl").getOrElse(false)
+
+      BasicAuthenticationFilterConfiguration(
+        string("realm").getOrElse(defaultRealm),
+        enabled,
```
Comment (simplesteph) [inline app/controllers/BasicAuthenticationFilter.scala]: @patelh @dpavlov  is this a bug, did you mean `ldapEnabled`? Or are we supposed to also set `basicAuthentication.enabled=true` to have LDAP?  Maybe the documentation could use some clarification here.

Comment (dpavlov) [reply app/controllers/BasicAuthenticationFilter.scala]: Yes you are right. LDAP works only with basic auth. Docs also have been changed to make it clear.

**Pull request merged**